File: app/src/main/java/com/nan/androidreview/designpattern/singleton/demo1/Singleton.java
package com.nan.androidreview.designpattern.singleton.demo1;

/**
 * 懒汉式
 * 特点：Lazy初始化；线程安全，但是由于每次需要同步性能较低，不建议使用
 */
public class Singleton {

    private static Singleton sInstance;

    private Singleton() {

    }

    public static synchronized Singleton getInstance() {
        if (sInstance == null) {
            sInstance = new Singleton();
        }
        return sInstance;
    }

}

File: app/src/main/java/com/nan/androidreview/designpattern/singleton/demo7/Singleton.kt
package com.nan.androidreview.designpattern.singleton.demo7

/**
 * Kotlin中通过object关键字可以实现最简单的单例
 * 特点：这种单例只有一个实现的对象；不能自定义构造方法；可以实现接口、继续父类
 */
object Singleton {
    public fun test() {
        println("")
    }
}

//调用方式
fun main(args: Array<String>) {
    Singleton.test()
}